cpp/src/arrow/dataset/file_base.cc

```diff
@@ -327,222 +322,70 @@ Status FileWriter::Finish() {
+    auto next_batch = groups.batches[index];
+    ARROW_ASSIGN_OR_RAISE(std::string destination,
+                          write_options.partitioning->Format(partition_expression));
+    (*counter)++;
+    return dataset_writer->WriteRecordBatch(next_batch, destination).Then([next_batch] {
+      return next_batch;
     });
-  }
-  return task_group->Finish();
+  };
+
+  return VisitAsyncGenerator(
+      std::move(partitioned_batch_gen),
+      [](const std::shared_ptr<RecordBatch>&) -> Status { return Status::OK(); });
 }
 
 }  // namespace
 
 Status FileSystemDataset::Write(const FileSystemDatasetWriteOptions& write_options,
                                 std::shared_ptr<Scanner> scanner) {
-  RETURN_NOT_OK(ValidateBasenameTemplate(write_options.basename_template));
-
-  // Things we'll un-lazy for the sake of simplicity, with the tradeoff they represent:
-  //
-  // - Fragment iteration. Keeping this lazy would allow us to start partitioning/writing
-  //   any fragments we have before waiting for discovery to complete. This isn't
-  //   currently implemented for FileSystemDataset anyway: ARROW-8613
-  //
-  // - ScanTask iteration. Keeping this lazy would save some unnecessary blocking when
-  //   writing Fragments which produce scan tasks slowly. No Fragments do this.
-  //
-  // NB: neither of these will have any impact whatsoever on the common case of writing
-  //     an in-memory table to disk.
-
-  ARROW_SUPPRESS_DEPRECATION_WARNING
-
-  // TODO(ARROW-11782/ARROW-12288) Remove calls to Scan()
-  ARROW_ASSIGN_OR_RAISE(auto scan_task_it, scanner->Scan());
-  ARROW_ASSIGN_OR_RAISE(ScanTaskVector scan_tasks, scan_task_it.ToVector());
-
-  ARROW_UNSUPPRESS_DEPRECATION_WARNING
-
-  WriteState state(write_options);
-  RETURN_NOT_OK(WriteInternal(*scanner->options(), &state, std::move(scan_tasks)));
-
-  auto task_group = scanner->options()->TaskGroup();
-  for (const auto& part_queue : state.queues) {
-    task_group->Append([&] {
-      RETURN_NOT_OK(write_options.writer_pre_finish(part_queue.second->writer().get()));
-      RETURN_NOT_OK(part_queue.second->writer()->Finish());
-      return write_options.writer_post_finish(part_queue.second->writer().get());
-    });
-  }
-  return task_group->Finish();
+  ARROW_ASSIGN_OR_RAISE(auto batch_gen, scanner->ScanBatchesAsync());
+  ARROW_ASSIGN_OR_RAISE(auto dataset_writer, DatasetWriter::Make(write_options));
+
+  AsyncGenerator<std::shared_ptr<int>> queued_batch_gen =
+      [batch_gen, &dataset_writer, &write_options]() -> Future<std::shared_ptr<int>> {
+    Future<TaggedRecordBatch> next_batch_fut = batch_gen();
+    return next_batch_fut.Then(
+        [&dataset_writer, &write_options](const TaggedRecordBatch& batch) {
+          if (IsIterationEnd(batch)) {
+            return AsyncGeneratorEnd<std::shared_ptr<int>>();
+          }
+          return WriteNextBatch(dataset_writer.get(), batch, write_options).Then([] {
+            return std::make_shared<int>(0);
```

Comment:
Just to make sure, the shared_ptr<int> is being used purely for the end-of-stream properties here?

Reply:
Ugh, yeah, I forgot about this bit of ugliness.  You are correct, I just needed something "iterable".  I've replaced this whole block with a task group in #11017 

https://github.com/apache/arrow/pull/11017/files#diff-2caf4e9bd3f139e05e55dca80725d8a9c436f5ccf65c76a37cebfa6ee9b36a6aR359-R367